### ISSUE TITLE
Bad throughput with high latency connections due to low TCP send buffer size

### ISSUE THREAD

**Issue by Medo42** [restic/restic#1996]
This is quite possibly related to #1383.

Output of `restic version`
--------------------------
restic 0.9.2 compiled with go1.10.3 on windows/amd64

How did you run restic exactly?
-------------------------------
restic.exe -r b2:bucket:/repo --verbose -o b2.connections=10 backup "D:\Eigene Dateien"

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Backblaze b2, but this should affect all backends with high latency using the http_transport.

Expected behavior
-----------------
restic should scan my folder and upload data to b2, trying its best to saturate my internet upload (~10MBit/s) since that should definitely be the bottleneck.

Actual behavior
---------------
restic scans my folder and uploads data to b2 at about 1.5MBit/s (~750KBit/s per actively used connection).

Steps to reproduce the behavior
-------------------------------
Here are environment details that may be relevant:
Windows 7 Professional, 64 bit
Zyxel NWD6605 USB wireless networking adapter
Location in Germany, Deutsche Telekom ISP

Do you have any idea what may have caused this?
-----------------------------------------------
First, I tried to see if the issue was with b2 or my internet connection instead of restic. But the python b2 command line client could upload a file and saturate my upload without an issue.
Next I fired up Wireshark to see on which end the data flow was restricted. It turned out that
- the bandwidth of each connection was quite steady
- there was no significant amount of dropped packets / retries, so probably not limited by congestion control
- the advertised window in the ACK packets coming back from b2 was sufficiently generous (around 300KB)
- however, when sending there were always just around ~12kB of data in flight. Given the high latency to the b2 servers, restic spent most of the time waiting for data to be ACKed by b2, and then immediately sent out a burst of new packets, then proceeded to wait again for enough outstanding bytes to be ACKed.

I jumped to the conclusion that restic was probably using a small TCP send buffer, since this would limit the ammount of outstanding bytes that the TCP stack could keep track of.

Do you have an idea how to solve the issue?
-------------------------------------------
To test the suspicion, I tried increasing the TCP send buffer size to 500KB (see https://github.com/Medo42/restic/commit/3c8bab2b42e638add86e139ec07d6cad12fc6bca) and it indeed made the situation a lot better, with the two active connections now taking most of my bandwidth. There is still quite a bit of fluctuation, but that could be due to other reasons.

Did restic help you or made you happy in any way?
-------------------------------------------------
Thank you so much for this project. I never saw a free/libre backup solution before that clicked with me. After using Crashplan for years in PC <-> PC mode I finally had to look for a new solution, and I was about ready to give my money to Backblaze when I saw restic mentioned on their blog. After watching the video of your talk at the C4 I was sure that this was the right tool for me. And I finally understand now how you can deduplicate data that's shifting around in files :smile: 

Note that I'm not sending this as a pull request but just include a proof of concept because I'm not sure this is actually a good general solution, or that this is the right place to put the code. I also went at this with zero knowledge of go, so adding this workaround was an interesting puzzle plus some trial and error, so I'm not sure if I got everything right.

If restic without my workaround never sets an explicit send buffer size, I don't really know how it is selected by the system, so this might just affect my version of Windows, or might be affected by some settings hidden behind arcane tools and registry keys.

**Comment by DavidWRobson** [comment]
I'm currently trying to back up about 10 TB containing ~20 million files to B2.  Even though my uplink is capable of 100MB/s, I typically only see ~1MB/s being transferred.  It looks like it will take ~6 weeks to do my initial upload.  If restic is using a small TCP send buffer size, this could well explain the poor performance I am seeing.

Is an increase in buffer size  likely to appear for version 0.9.3?    Having the size as an option would be nice.

**Comment by mholt** [comment]
@Medo42 This looks like a good change -- I don't know a whole lot about implications of a larger buffer (other than more memory usage with lots of concurrent connections, which I don't think we have here?) but I think you should submit a pull request. :) I would also be happy to see faster transfer speeds, especially if the fix is this simple.

**Comment by Medo42** [comment]
One reason I didn't is that my workaround doesn't feel like a clean solution, because it just pulls the value of 500k out of nowhere (and it would be too small to e.g. exhaust the connection of @DavidWRobson). Making it configurable goes against what I understand as a design principle of restic, which is to do the right thing automatically instead of offering lots of configuration options. Since you don't have to fiddle with this setting on other software which does large uploads (e.g. your browser), I assume there is a better solution. However, until we figure it out, the workaround is definitely a big improvement, so I'll see what I can do.

**Comment by Medo42** [comment]
I added a commandline option (--tcp-sendbuffer-size) to my workaround version now. Here is a Windows/amd64 binary for you to play around with if you dare: https://github.com/Medo42/restic/releases/tag/v0.9.2.sendbuffersize

**Comment by DavidWRobson** [comment]
Thanks!  Can I be cheeky and ask for a Linux x86_64 binary?

I think that my performance issues may be a combination of this issue and also issue 1976, as I too am running on GPFS which would benefit from o.FileReadConcurrency.  Again, a command line option would be great.

I understand that having options is against restic's design principle, but I don't understand how restic  is going to detect the environment that it is running in (e,g, network speed, filesystem type and performance) in order to configure itself optimally.

Maybe have restic to optimize itself for a typical environment (what ever that means), but still have options for this behaviour to be over written, as the gains may be large.  I'm sure that, in my case, being able to configure tcp buffer sizes and the number of concurrent files may double or triple performance.

Also, I would like to say that restic is awesome software!

**Comment by Medo42** [comment]
@DavidWRobson Hi, I finally got around to building that Linux binary you asked for. See this release: https://github.com/Medo42/restic/releases/tag/v0.9.4.sendbuffersize

I'd be interested to know if this fixes the problem for you as well. Some comparisons with/without expanded buffer size would be helpful.

**Comment by DavidWRobson** [comment]
@Medo42 Thanks very much for doing that.

I've run six backup runs with the vanilla 0.9.4 client against a backblaze repository, and then run six similar runs, with your client with a buffer size of 15000, each time using iotop to measure the https traffic to backblaze.  Your client ran ~ 15% faster.  I was hoping for more, as my bandwidth is nowhere near filled.  However 15% faster is significant, so again, thanks very much.

Is your change likely to make it into the main branch?

**Comment by Medo42** [comment]
Hmm, for me it was more like a 5x difference, going from 1.5MBit/s with default settings up to ~8MBit/s with larger send buffer. I'd also guess that the Linux networking stack might behave differently from the Windows one in general. So I'd be cautious in concluding that we're actually seeing the same problem at all.

What are the absolute speeds that you see? How does that compare to uploading a file with Backblaze's official b2 client? The official client saturated my connection, which is how I realized in the first place that the issue must be with restic.

**Comment by DavidWRobson** [comment]
Thanks for suggesting the b2 official client.  The version I had built  a while back  doesn't work on my file server (where I am running restic) because of glibc issues, so instead of rebuilding, I first running the b2 client on my NFS clients.  I got speeds of ~100MB/s.  I then tried running restic on my NFS clients and was surprised to find that, despite the extra NFS overhead, I again got speeds of ~100MB/s.

So ...
My NFS clients can talk to the NFS server and the internet at 100MB/s
My NFS server can talk to the NFS clients at 100MB/s, but not to the internet.

So my "restic" problem is, in fact, an internal network performance problem.  I will get our network team to investigate.

Sorry for wasting your time, but I did see an improvement with your patch, and I am sure it will almost  benefit everyone.

**Comment by fd0** [comment]
Hey, sorry for not coming back to you earlier. I'm just going through all the unlabeled issues in the restic repo now that I have a bit more spare time.

B2 is a bit peculiar, especially for non-US users: the latency from Europe to B2's servers is very high, at least 800ms from "HTTP headers sent" to "HTTP response header received", so the per-file overhead is very high.

Most files uploaded to B2 are rather small, especially compared to high-bandwidth uplinks: by default, most pack files restic creates and uploads to the backend are ~4MiB in size. This will also limit throughput, as it's much more efficient to upload a small number of large files than a large number of small files. Our plan is to adapt the file size dynamically to how much upstream bandwidth (and memory) is available at backup time, but we're not there yet.

I'm wondering though, why didn't the OS set a good TCP send buffer itself? Isn't that automatically set? Sure, we can increase the send buffer size for the B2 backend, ideally so that most files will just fit into the buffer. Maybe 6 MiB is a good size?

Btw, in https://github.com/Medo42/restic/commit/3c8bab2b42e638add86e139ec07d6cad12fc6bca you're setting 5MiB as the TCP send buffer size, not 500KiB. :)

Before we can merge such a change, I'd love to see some benchmarks and comparisons, so it'd be great to build a small sample program that measures the effects which users can run on different operating systems.

**Comment by Medo42** [comment]
It seems surprisingly hard to find any official information about this since it's such a "core" feature of the operating system, but from what I could find out with some googling it appears that Windows Vista and 7 default to an 8KB send buffer, and both older and newer versions behave better: https://osqa-ask.wireshark.org/questions/57810/slow-transfer-on-high-latency-high-bandwidth-link

This *might* only affect socket operations in nonblocking mode. Various workarounds are suggested, including setting a higher default sendbuffer size through an undocumented registry parameter and enabling a dynamic mode for the sendbuffer with another registry param. I'll try whether those fix the issue for me, but it's probably a good idea for restic to offer good performance without having to twiddle with the registry first.

I can also look into creating the test program, although it might take me a bit to get around to it.

**Comment by Medo42** [comment]
I tried changing the registry values indicated in the article I linked above, and then running backups with official restic build 0.9.4 with rclone backend.

First, I tried again on the unchanged system and got the same result as before, i.e. a fairly slow upload speed of ~1.5Mbit/s.

I apparently had the following registry value set already, probably from trying to follow advice from the internet:
`HKLM\SYSTEM\CurrentControlSet\services\AFD\Parameters\DynamicSendBufferDisable (DWORD) = 0`
I removed that value and instead added DefaultSendWindow (DWORD) in the same key using a buffer size of 512K. In other words:
`HKLM\SYSTEM\CurrentControlSet\services\AFD\Parameters\DefaultSendWindow (DWORD) = 524288`
After a reboot, I tested again and saw the same behaviour as when setting the sendbuffer size manually within restic, i.e. the upload speed increased to something like 7-8 MBit/s.

So with this change I can finally use the official restic build again without performance issues :)

**tl;dr:**
If you are running Windows Vista or Windows 7 and you're experiencing slow upload to b2, you can try the following workaround / system tuning:
- Open regedit
- navigate to the key HKLM\SYSTEM\CurrentControlSet\services\AFD\Parameters
- Add a DWORD value "DefaultSendWindow"
- Set the input to "decimal" and enter 524288 (=512KiB)
- reboot and try running your backup again

If you see a significant increase in upload speed, then you can try if even larger values help more (or calculate a good buffer size for your connection). If you don't see a significant difference then your problem is probably something else and you may want to delete the registry value again.

**Comment by frittentheke** [comment]
@fd0  I did build restic 0.9.5 with the commits from @Medo42 implementing the configurable tcp-sendbuffer-size and attempted some comparisons when sending data to Backblaze B2 ... while I believe (!) it makes a difference in upload speed, even with a big send buffer of 8192 there still are drops in data rate. Regarding my test data, this appears to be the case especially for lots of small files.

Especially with B2 it seems doing multiple concurrent connections would be most beneficial to keep ones own internet uplink busy - just like they suggest themselves https://help.backblaze.com/hc/en-us/articles/115002880514

**Comment by Medo42** [comment]
@frittentheke For reference, what is your OS and location / ping to the b2 servers?

I can confirm that the upload speed is not very steady, so raising the number of connections may be worthwhile as well if that leads to a further improvement. But first, I don't think your linked article counts as independend information from Backblaze that using more connections is better, because it apparently simply copied that advice from https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html#backblaze-b2 . Second, while more concurrent connections to b2 would also help to saturate the bandwidth, I think there are two separate causes for bad throughput here and imo it shouldn't be a question of fixing either one or the other :)

**Comment by frittentheke** [comment]
@Medo42 I am actually on Linux (Arch + Mint 19.2) and pushing my bits out from Germany.

**Comment by macropin** [comment]
Performance from a FreeBSD host in Australia to b2 US.
```
[111:00:44] 26.16%  6828 files 53.315 GiB, total 26102 files 203.827 GiB, 0 errors ETA 313:23:49
```
I've been following this since June, hoping to see some movement... If anyone has patches they want tested, send them my way.

**Comment by arichiardi** [comment]
I apologize in advance for not contributing more than just a data point here but I am also seeing very slow upload speeds

```
scan finished in 1.917s: 2 files, 4.379 GiB
modified  /var/lib/dir/file.enc.md5, saved in 0.001s (44 B added)
modified  /var/lib/dir/file.enc, saved in 841.057s (21.533 MiB added)
modified  /var/lib/dir/, saved in 841.073s (0 B added, 194.604 KiB metadata)
modified  /var/dir/, saved in 841.073s (0 B added, 369 B metadata)
modified  /var/, saved in 841.074s (0 B added, 367 B metadata)

Files:           0 new,     2 changed,     0 unmodified
Dirs:            0 new,     3 changed,     0 unmodified
Data Blobs:     12 new
Tree Blobs:      4 new
Added to the repo: 21.725 MiB
```

It seems 22 MB were uploaded in 840 seconds to Backblaze, which is pretty bad.

Please let me know what I can try out, thanks a lot!

**Comment by Medo42** [comment]
@arichiardi I think you might want to open a separate issue for that, I doubt that this problem is related to the issue I was having which (to the best of my knowledge) only affects Windows 7 and Vista. Since both of those are fairly obsolete by now, I'll close this issue.

**Comment by joachimneu** [comment]
Posting this here in case someone else gets to this issue from some search engine:

Located on the US west coast, using restic to backup to Backblaze's central EU location (which I believe is in Amsterdam? so probably around 150ms ping). *Decreasing* the number of B2 connections from the default of 5 to 2 improved throughput from ~2MB/s to ~12MB/s.
